torch/distributed/checkpoint/state_dict_saver.py

```diff
@@ -105,21 +107,47 @@ def save(
     """
     torch._C._log_api_usage_once("torch.distributed.checkpoint.save")
 
-    dumpable_state_dict = {}
-    for key, elem in state_dict.items():
-        dumpable_state_dict[key] = (
-            elem.state_dict() if isinstance(elem, Stateful) else elem
-        )
-
     return _save_state_dict(
-        dumpable_state_dict,
+        _stateful(state_dict),
         storage_writer,
         process_group,
         coordinator_rank,
         no_dist,
         planner,
     )
 
+def async_save(state_dict: STATE_DICT_TYPE, *args: Any, **kwargs: Any) -> Future:
+    """Asynchronous version of ``save_state_dict``. This code first de-stages the state_dict on CPU, and then calls
+    `save` in a separate thread.
+
+    .. warning::
+        This feature is experimental and subject to removal/change.
+
+    Returns:
+        Future: A future holding the resultant Metadata object from `save`.
+
+    """
+
+    # since calling nccl collectives in new threads is risky, we call `_stateful` here in case users have
+    # defined custom collectives
```

Comment:
I don't quite understand this comment. Where are those custom collectives from?

Reply:
Users can implement custom collectives in `state_dict` and `load_state_dict`

Reply:
hmm, but this comment is note quite right. Even if there are no NCCL collectives in the state_dict and load_state_dict, we still have to call state_dict() and move it to CPU before we pass the state_dict to the saver thread. I prefer removing this comment or change it to explain the correct reason.

Reply:
Correct. We would have to do it in this order regardless, great point! Thanks @fegin